subsys/bluetooth/host/conn.c

```diff
@@ -513,6 +546,11 @@ static bool send_frag(struct bt_conn *conn, struct net_buf *buf, uint8_t flags,
 	if (IS_ENABLED(CONFIG_BT_ISO) && conn->type == BT_CONN_TYPE_ISO) {
 		err = send_iso(conn, buf, flags);
 	} else if (IS_ENABLED(CONFIG_BT_CONN)) {
+		/* Buffer user data may be overridden by send_acl().
+		 * Its lifetime is no longer tied to the buffer, and freeing it
+		 * must be handled separately.
+		 */
+		tx_data(buf)->tx = NULL;
```

Comment:
This part here isn't clear to me. Can you elaborate?

Reply:
In `send_acl()` `bt_buf_set_type(buf, BT_BUF_ACL_OUT);` is called unless an invalid flag is given, which overrides the buffer user data. Later on when `acl_tx_pool->_destroy` is called, the metadata will not be freed because the user data does not contain a valid pointer, but rather the value of `BT_BUF_ACL_OUT`. If `bt_buf_set_type()` is not called, the user data is a valid pointer and we can end up with a double-free.

Reply:
The conn tx meta data is added to the `conn->tx_pending` list in the success case and the callbacks will be called when the num_complete HCI event happens. The buffer is unref-ed by the HCI driver when it has been sent to the controller. In the failure case the buffer is unref-ed in `bt_conn_process_tx`.

Reply:
Alright. 

But what happens with the data that were in `tx_data(buf)->tx` originally?

Reply:
The data in the user buffer is just a pointer to the metadata, so the metadata itself is not overridden. We have another pointer to the metadata `tx`, set at the top of `send_frag`.

We should probably reset the user data in the failure case, so that it will be freed when `net_buf_unref` is called. Will test.

Reply:
If user data is not reset here a bunch of tests fail. It may not be a double-free, but for some reason a pointer to some invalid data is added to `conn->tx_complete` and the test segfaults when it tries to call a function that does not exist.

Reply:
Figured it out. `acl_send()` calls `bt_buf_set_type()`, which sets the first byte of the user data to 2 and leaves the rest as is. Later on when `net_buf_unref()` and `_destroy()` are called, the user data is interpreted as a pointer. Because the system is little-endian only the least significant byte of the pointer has been changed. `PART_OF_ARRAY(conn_tx, tx)` still gives `true` even though `tx` is not an element in `conn_tx`, just between the bounds. `tx_free()` writes to `tx` and zeroes out parts of two elements of `conn_tx`. Later, when the callback is called, the callback pointer is all mangled and we segfault.

Reply:
Maybe `bt_buf_set_type()` should zero out the parts of the user data that is not used?

Reply:
> Because the system is little-endian only the least significant byte of the pointer has been changed. 

I hope that the code does not assume any endianess - Zephyr runs on both big and little endian systems.

Reply:
> Maybe `bt_buf_set_type()` should zero out the parts of the user data that is not used?

I wish I could help here, but I simply do not understand the issue nor the context well enough. @jhedberg?

Reply:
>> Because the system is little-endian only the least significant byte of the pointer has been changed.
>
>I hope that the code does not assume any endianess - Zephyr runs on both big and little endian systems.

On a big-endian system I believe the most significant byte would be overridden instead and the pointer would point outside of the `conn_tx` array and `PART_OF_ARRAY` would give `false` and `tx` thus not being mangled.